synapse/rest/client/login.py

```diff
@@ -172,7 +172,12 @@ async def on_POST(self, request: SynapseRequest) -> Tuple[int, LoginResponse]:
 
         try:
             if login_submission["type"] == LoginRestServlet.APPSERVICE_TYPE:
-                appservice = self.auth.get_appservice_by_req(request)
+                requester = await self.auth.get_user_by_req(request)
+                appservice = requester.app_service
+
+                if appservice is None:
+                    # This is not an application service
+                    raise InvalidClientTokenError()
```

Comment:
```suggestion
                    raise InvalidClientTokenError("This login method is only valid for application services")
```

am I missing a reason why you didn't do this?

Reply:
Not really 🤷‍♂️